File: Java/java_labwork/14-11-25/Appliance.java
package lab;

abstract class Appliance
{
	String brand;
	double price;
	int powerRating; 
	
	Appliance() // default constructor
	{
		this.brand = "Samsung";
		this.price = 25000;
		this.powerRating = 1200;
	}
	
	Appliance(String brand, double price, int powerRating) // parameterized constructor
	{
		this.brand = brand;
		this.price = price;
		this.powerRating = powerRating;
	}
	
	// getter and setter 
	public String getBrand() {
		return brand;
	}
	public void setBrand(String brand) {
		this.brand = brand;
	}
	public double getPrice() {
		return price;
	}
	public void setPrice(double price) {
		this.price = price;
	}
	public int getPowerRating() {
		return powerRating;
	}
	public void setPowerRating(int powerRating) {
		this.powerRating = powerRating;
	}
	
	abstract void start();
	
	public String toString() 
	{
	    return "Brand: " + this.brand +" Price: " + this.price +" Power Rating: " + this.powerRating + "W";
	}

	
} // class Appliance ends here


class WashingMachine extends Appliance
{
	int drumSize; 
	
	WashingMachine() // default constructor
	{
		super();
		this.drumSize = 7;
	}
	
	WashingMachine(String brand, double price, int powerRating, int drumSize) // parameterized constructor
	{
		super(brand, price, powerRating);
		this.drumSize = drumSize;
	}
	
	// getter and setter 
	public int getDrumSize() {
		return drumSize;
	}
	public void setDrumSize(int drumSize) {
		this.drumSize = drumSize;
	}
	
	void start()
	{
		System.out.println("Washing machine started intakeing water");
	}
	
	public String toString()
	{
		
		return super.toString()+"Drum Size: " + this.drumSize + "kg";
	}
	
} // class WashingMachine ends here


class Refrigerator extends Appliance
{
	int capacity; 
	
	Refrigerator() // default constructor
	{
		super();
		this.capacity = 300;
	}
	
	Refrigerator(String brand, double price, int powerRating, int capacity) // parameterized constructor
	{
		super(brand, price, powerRating);
		this.capacity = capacity;
	}
	
	// getter and setter 
	public int getCapacity() {
		return capacity;
	}
	public void setCapacity(int capacity) {
		this.capacity = capacity;
	}
	
	void start()
	{
		System.out.println("Refrigerator started cooling!");
	}
	
	public String toString()
	{
		
		return super.toString()+"Capacity: " + this.capacity + "L" ;
	}
	
	void display()
	{
		System.out.println();
		System.out.println();
	}
	
} // class Refrigerator ends here


class Microwave extends Appliance
{
	int volume; 
	
	Microwave() // default constructor
	{
		super();
		this.volume = 25;
	}
	
	Microwave(String brand, double price, int powerRating, int volume) // parameterized constructor
	{
		super(brand, price, powerRating);
		this.volume = volume;
	}
	
	// getter and setter 
	public int getVolume() {
		return volume;
	}
	public void setVolume(int volume) {
		this.volume = volume;
	}
	
	void start()
	{
		System.out.println("Microwave started heating food!");
	}
	
	public String toString()
	{
		
		return super.toString()+"Volume: " + this.volume + "L";
	}
	
} // class Microwave ends here


class ApplianceTest
{
	public static void main(String[] args)
	{
		Appliance a1;
		
		System.out.println("--- WashingMachine ---");
		 a1 = new WashingMachine("Whirlpool", 32000, 1500, 8);
		 System.out.println(a1.toString());
		 
		 System.out.println("--- Refrigerator ---");
		 a1 = new Refrigerator("Godrej", 40000, 1300, 350);
		 System.out.println(a1.toString());
		 
		 System.out.println("--- Microwave ---");
		 a1 = new Microwave("Panasonic", 15000, 800, 30);
		 System.out.println(a1.toString());
		 
		
	} // main ends here
	
} // class ApplianceTest ends here

File: c++/Assignment16/Polymorphism by structure/Appliance.cpp
#include <iostream>
#include <string>
using namespace std;

// -------------------- Appliance --------------------
struct Appliance
{
protected:
    string brand;
    double price;
    int powerRating;

public:
    // default constructor
    Appliance()
    {
        brand = "Samsung";
        price = 25000;
        powerRating = 1200;
    }

    // parameterized constructor
    Appliance(string b, double p, int pr)
    {
        brand = b;
        price = p;
        powerRating = pr;
    }

    virtual void start()
    {
        cout << "Appliance started!" << endl;
    }

    virtual void display()
    {
        cout << "Brand: " << brand << endl;
        cout << "Price: " << price << endl;
        cout << "Power Rating: " << powerRating << "W" << endl;
    }
};

// -------------------- WashingMachine --------------------
struct WashingMachine : public Appliance
{
protected:
    int drumSize;

public:
    WashingMachine() : Appliance()
    {
        drumSize = 7;
    }

    WashingMachine(string b, double p, int pr, int d)
        : Appliance(b, p, pr)
    {
        drumSize = d;
    }

    void start()
    {
        cout << "Washing machine started intakeing water" << endl;
    }

    void display()
    {
        Appliance::display();
        cout << "Drum Size: " << drumSize << "kg" << endl << endl;
    }
};

// -------------------- Refrigerator --------------------
struct Refrigerator : public Appliance
{
protected:
    int capacity;

public:
    Refrigerator() : Appliance()
    {
        capacity = 300;
    }

    Refrigerator(string b, double p, int pr, int c)
        : Appliance(b, p, pr)
    {
        capacity = c;
    }

    void start()
    {
        cout << "Refrigerator started cooling!" << endl;
    }

    void display()
    {
        Appliance::display();
        cout << "Capacity: " << capacity << "L" << endl << endl;
    }
};

// -------------------- Microwave --------------------
struct Microwave : public Appliance
{
protected:
    int volume;

public:
    Microwave() : Appliance()
    {
        volume = 25;
    }

    Microwave(string b, double p, int pr, int v)
        : Appliance(b, p, pr)
    {
        volume = v;
    }

    void start()
    {
        cout << "Microwave started heating food!" << endl;
    }

    void display()
    {
        Appliance::display();
        cout << "Volume: " << volume << "L" << endl << endl;
    }
};

// -------------------- ApplianceTest --------------------
int main()
{
    Appliance* a1;

    cout << "--- Appliance ---" << endl;
    a1 = new Appliance("LG", 18000, 1000);
    a1->display();
    a1->start();

    cout << "\n--- WashingMachine ---" << endl;
    a1 = new WashingMachine("Whirlpool", 32000, 1500, 8);
    a1->display();
    a1->start();

    cout << "\n--- Refrigerator ---" << endl;
    a1 = new Refrigerator("Godrej", 40000, 1300, 350);
    a1->display();
    a1->start();

    cout << "\n--- Microwave ---" << endl;
    a1 = new Microwave("Panasonic", 15000, 800, 30);
    a1->display();
    a1->start();
}

